### ISSUE TITLE
Upgrade the YOLO model with features from new variants

### ISSUE THREAD

**Issue by senarvi** [Lightning-AI/lightning-bolts#816]
## 🚀 Feature

Several new variants of the YOLO model have been proposed in the past couple of years. The current implementation in PyTorch Lightning Bolts is based on the original Darknet implementation. I have updated it with features from the new PyTorch based variants and created a pull request.

### Motivation

The code has been refactored, allowing easy variation of the network architecture, target matching algorithm, or loss function. Some important changes that have been proposed to these algorithms are implemented in this pull request in a modular way. This is the only implementation that combines features from the latest variants, including YOLOv5 and YOLOX, as well as allows loading Darknet based models.

You can find my pull request here: https://github.com/PyTorchLightning/lightning-bolts/pull/817

**Comment by senarvi** [comment]
This PR also includes the fix to the number of input channels by @kfirgedal . Maybe you could look at this pull request and say what you think?

**Comment by senarvi** [comment]
I added the deeper P6 variant of YOLOv4. The depth of the network and the width of each stage can also be configured freely.

I think there are pretty significant improvements in the pull request. Is anyone interested in having a look? @kfirgedal @Borda @ethanwharris @awaelchli

**Comment by senarvi** [comment]
I did still some minor refactoring, but otherwise this pull request is pretty stable already. I think there was interest in merging this. @luca-medeiros do you have time to look at this?
